File: frontend/src/types/booking.ts
export type SeatStatus = "AVAILABLE" | "LOCKED" | "SOLD";
export type BookingStatus = "PENDING" | "CONFIRMED" | "EXPIRED" | "CANCELLED";
export type PaymentStatus = "PENDING" | "COMPLETED" | "FAILED";
export type TicketStatus = "VALID" | "USED" | "CANCELLED";

export interface SeatWithZone {
  id: string;
  event_id: string;
  zone_id: string | null;
  zone_name: string | null;
  zone_color: string | null;
  price: string | null;
  row_number: number;
  seat_number: number;
  status: SeatStatus;
  locked_by: string | null;
}

export interface SeatLockResponse {
  seat_id: string;
  status: SeatStatus;
  expires_in: number;
}

export interface BookingItem {
  id: string;
  seat_id: string;
  row_number: number;
  seat_number: number;
  zone_name: string;
  price: string;
}

export interface Payment {
  id: string;
  amount: string;
  method: string;
  status: PaymentStatus;
  paid_at: string | null;
}

export interface Booking {
  id: string;
  event_id: string;
  event_title: string;
  event_date: string;
  status: BookingStatus;
  total_amount: string;
  created_at: string;
  expires_at: string;
  items: BookingItem[];
  payment: Payment | null;
}

export interface Ticket {
  id: string;
  booking_item_id: string;
  qr_data: string;
  qr_image: string;
  status: TicketStatus;
  issued_at: string;
  event_title: string;
  event_date: string;
  zone_name: string;
  row_number: number;
  seat_number: number;
}

export interface SeatUpdateMessage {
  type: "seat_update";
  seat_id: string;
  zone_id: string | null;
  row_number: number;
  seat_number: number;
  status: SeatStatus;
  locked_by: string | null;
}

File: backend/app/schemas/booking.py
from datetime import datetime
from decimal import Decimal
from uuid import UUID

from pydantic import BaseModel, Field

from app.utils.enums import BookingStatus, PaymentStatus, TicketStatus


class BookingCreate(BaseModel):
    event_id: UUID
    seat_ids: list[UUID] = Field(min_length=1, max_length=20)


class BookingItemOut(BaseModel):
    id: UUID
    seat_id: UUID
    row_number: int
    seat_number: int
    zone_name: str
    price: Decimal

    class Config:
        from_attributes = True


class PaymentOut(BaseModel):
    id: UUID
    amount: Decimal
    method: str
    status: PaymentStatus
    paid_at: datetime | None

    class Config:
        from_attributes = True


class BookingOut(BaseModel):
    id: UUID
    event_id: UUID
    event_title: str
    event_date: datetime
    status: BookingStatus
    total_amount: Decimal
    created_at: datetime
    expires_at: datetime
    items: list[BookingItemOut]
    payment: PaymentOut | None


class TicketOut(BaseModel):
    id: UUID
    booking_item_id: UUID
    qr_data: str
    qr_image: str  # data-URL PNG
    status: TicketStatus
    issued_at: datetime
    event_id: UUID
    event_title: str
    event_status: str
    event_date: datetime
    venue_name: str
    venue_address: str
    price: Decimal
    zone_name: str
    row_number: int
    seat_number: int
